File: apps/chuzz/src/toolbar.rs
//! Back, forward, reload, and the address bar.

use dioxus_native::prelude::*;

use crate::nav::request_from_input;
use crate::tab::{Tab, TabId, TabStoreImplExt, active_tab};

fn is_enter(key: &Key) -> bool {
    matches!(key, Key::Enter) || matches!(key, Key::Character(text) if text == "\n")
}

#[component]
pub fn Toolbar(
    tabs: Store<Vec<Tab>>,
    active_tab_id: Signal<TabId>,
    mut url_input_value: Signal<String>,
    mut focus_address_bar: Signal<bool>,
) -> Element {
    // The address bar's element handle, captured when it mounts. A shortcut can
    // only ask for focus by name; something has to hold the node it names.
    let mut address_bar: Signal<Option<std::rc::Rc<MountedData>>> = use_signal(|| None);
    use_effect(move || {
        if !focus_address_bar() {
            return;
        }
        if let Some(element) = address_bar() {
            spawn(async move {
                let _ = element.set_focus(true).await;
            });
        }
        // Cleared whether or not the element was there: the signal is a request,
        // not a state, and leaving it set would make the next request a no-op
        // because the effect only re-runs on a change.
        focus_address_bar.set(false);
    });
    // The address bar follows the active tab, but only when that tab actually
    // navigates. Tracking the URL itself is what makes typing possible: an
    // effect that re-ran on every render would overwrite each keystroke with
    // the current URL, and the field could never hold anything typed.
    let current_url = active_tab(tabs, active_tab_id()).current_url().to_string();
    use_effect(use_reactive!(|current_url| {
        *url_input_value.write_unchecked() = current_url;
    }));

    let tab = active_tab(tabs, active_tab_id());
    let can_go_back = tab.can_go_back();
    let can_go_forward = tab.can_go_forward();

    let submit = use_callback(move |_| {
        let typed = url_input_value.read().clone();
        if let Some(req) = request_from_input(&typed) {
            active_tab(tabs, active_tab_id()).navigate(req);
        }
    });

    let back_class = if can_go_back {
        "tool-button"
    } else {
        "tool-button disabled"
    };
    let forward_class = if can_go_forward {
        "tool-button"
    } else {
        "tool-button disabled"
    };

    rsx!(
        div { id: "toolbar",
            div {
                class: "{back_class}",
                title: "Back",
                onclick: move |_| active_tab(tabs, active_tab_id()).go_back(),
                "\u{2190}"
            }
            div {
                class: "{forward_class}",
                title: "Forward",
                onclick: move |_| active_tab(tabs, active_tab_id()).go_forward(),
                "\u{2192}"
            }
            div {
                class: "tool-button",
                title: "Reload",
                onclick: move |_| active_tab(tabs, active_tab_id()).reload(),
                "\u{21bb}"
            }
            input {
                id: "url-bar",
                r#type: "text",
                value: "{url_input_value}",
                placeholder: "Search or enter address",
                onmounted: move |event| address_bar.set(Some(event.data())),
                oninput: move |event| url_input_value.set(event.value()),
                onkeydown: move |event| {
                    if is_enter(&event.key()) {
                        submit.call(());
                    }
                },
            }
        }
    )
}

File: apps/chuzz/frontend/src/features/toolbar/toolbar.ts
import { createEffect, createSignal } from "solid-js";
import { useBrowser } from "~/stores/browser";

/**
 * The address field's behaviour. No JSX, no class names.
 *
 * Keeps the one subtlety the Rust original had: the field tracks the active
 * tab's URL, but only when that URL actually changes. An effect that re-ran on
 * every render would overwrite each keystroke with the current address, and
 * the field could never hold anything typed.
 */
export function createToolbar() {
  const browser = useBrowser();
  const [typed, setTyped] = createSignal("");
  const [refused, setRefused] = createSignal(false);

  createEffect((previous: string | undefined) => {
    const url = browser.activeTab()?.url ?? "";
    if (url !== previous) {
      setTyped(url);
      setRefused(false);
    }
    return url;
  });

  return {
    typed,
    // A `state` key of the recipe, so it reaches the class map by name.
    refused,
    canGoBack: () => Boolean(browser.activeTab()?.canGoBack),
    canGoForward: () => Boolean(browser.activeTab()?.canGoForward),
    goBack: () => browser.goBack(),
    goForward: () => browser.goForward(),
    reload: () => browser.reload(),
    setTyped: (value: string) => {
      setTyped(value);
      setRefused(false);
    },
    submit: async () => {
      // `nav.rs` decides what a typed string means. A value that is neither a
      // URL nor a bare hostname is not a search, and the shell says so rather
      // than navigating somewhere unrelated; showing that beats looking broken.
      const accepted = await browser.navigate(typed());
      setRefused(!accepted);
    },
  };
}

export type ToolbarModel = ReturnType<typeof createToolbar>;
